code/obj/flock/structure/sapper.dm

```diff
@@ -0,0 +1,147 @@
+#define CHARGING_BITS "Speed up Flockbits"
+#define CHARGING_DRONES "Charge incapacitors"
+#define CHARGING_STRUCTURES "Enhance structures"
+
+/obj/flock_structure/sapper
+	icon_state = "sapper-off" // current sprites are placeholders
+	name = "sparking machine"
+	desc = "A strange structure supporting four coils. You can feel the electricity coming off of it"
+	flock_desc = "A multipurpose support structure that saps power from a local APC. It can be set to enchance Flockbits, Flockdrones, or structures."
+	flock_id = "Sapper"
+	health = 30
+	health_max = 30
+	repair_per_resource = 1.5
+	resourcecost = 100
+
+	passthrough = TRUE
+
+	var/mode = CHARGING_BITS
+	var/mode_cooldowns = list(CHARGING_BITS = 20 SECONDS, CHARGING_DRONES = 30 SECONDS, CHARGING_STRUCTURES = 30 SECONDS)
+
+	var/obj/machinery/power/apc/linked_apc = null
+
+	New(atom/location, datum/flock/F = null)
+		..()
+		src.linked_apc = src.try_link_apc()
+		if (!src.linked_apc)
+			src.icon_state = "sapper-off"
+
+		src.info_tag.set_info_tag("Mode: [src.mode]")
+
+		ON_COOLDOWN(src, CHARGING_BITS, src.mode_cooldowns[CHARGING_BITS])
+		ON_COOLDOWN(src, CHARGING_DRONES, src.mode_cooldowns[CHARGING_DRONES])
+		ON_COOLDOWN(src, CHARGING_STRUCTURES, src.mode_cooldowns[CHARGING_STRUCTURES])
+
+	building_specific_info()
+		return {"<span class='bold'>Mode:</span> [src.mode].
+				<br><span class='bold'>Linked power supply charge:</span> [src.linked_apc?.cell ? "[round(src.linked_apc.cell.charge / src.linked_apc.cell.maxcharge * 100)]%": "Not linked"]."}
+
+	process(mult)
+		if (!src.linked_apc)
+			src.try_link_apc()
+			if (!src.linked_apc)
+				src.icon_state = "sapper-off"
+				return
```

Comment:
Since `try_link_apc` returns the APC if it found one and `null` otherwise, this is a little prettier
```suggestion
		if (!src.linked_apc && !src.try_link_apc())
			src.icon_state = "sapper-off"
			return
```